File: algorithms/Ruby/ShellSort/ShellSort.rb
#Shell Sort implementation(Diminishing Increment Sort)
#Time-complexity: O(n^2), In-place
#will be using Knuth series :3n+1

def shell_sort(a)
    n=a.length
    h=1
    
    while (h<n/3)  #for computing increment factor "h"
        h= (3*h)+1
    end
    
    while h>=1
    # Logic of insertion sort with inrement steps of "h"
        for i in h...n
            j=i
            while j>=h
                if a[j-h]>a[j]
                    temp=a[j]
                    a[j]=a[j-h]
                    a[j-h]=temp
                end
                j-=h
            end
        end
        h/=3
    end
    
    return a
    
end

puts(shell_sort([0,5,4,7,1,8,9,3,7,1,4,2,8,6]))

File: algorithms/sorting/shell-sort/swift/ShellSort.swift
class ShellSort {
    static func sort(_ arr: inout [Int]) {
        let n = arr.count
        var gap = n / 2
        
        while gap > 0 {
            for i in gap..<n {
                let temp = arr[i]
                var j = i
                while j >= gap && arr[j - gap] > temp {
                    arr[j] = arr[j - gap]
                    j -= gap
                }
                arr[j] = temp
            }
            gap /= 2
        }
    }
}
